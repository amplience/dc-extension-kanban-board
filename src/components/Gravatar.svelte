<script>
  import type { User } from 'dc-extensions-sdk/dist/types/lib/components/Users';
  import { PRESETS } from '../services/models/status';
  import { gravatars } from '../services/stores/gravatar';
  import md5 from 'md5';


  const GRAVATAR_URL = 'https://www.gravatar.com/avatar' as const;
  const GRAVATAR_QUERY = '?d=404&s=56' as const;
  const GRAVATAR_COLORS = Object.keys(PRESETS);

  export let user: User;

  const emailHash = md5(user.email);

  async function resolveGravatar(): Promise<string | null> {
    let gravatarUrl = $gravatars.get(emailHash);
    if (gravatarUrl !== undefined) {
      return gravatarUrl
    }

    gravatarUrl = `${GRAVATAR_URL}/${emailHash}${GRAVATAR_QUERY}`;
    const res = await fetch(gravatarUrl);
    if(!res.ok) {
      gravatarUrl = null;
    }

    $gravatars.set(emailHash, gravatarUrl);
    return gravatarUrl;
  }

  function getUserInitial(): string {
    return user.firstName.charAt(0).toUpperCase();
  }

  function getColor(): string {
    if (!emailHash) {
      return '#999';
    }
    const numFromHash = parseInt(emailHash.replace(/[^\d]+/, ''), 10);
    return GRAVATAR_COLORS[numFromHash % GRAVATAR_COLORS.length];
  }
</script>

<style lang="scss">
  .gravatar {
    display: inline-flex;
    width: 28px;
    height: 28px;
    border-radius: 100%;
    overflow: hidden;
    color: #fff;
    text-transform: uppercase;
    margin: 5px 0;
    font-weight: 500;
    font-size: 16px;
  }

  .gravatar img {
    width: 28px;
    height: 28px;
  }

  .gravatar-intial {
    background-color: var(--gravatar-color);
    display: flex;
    flex: 1 1 100%;
    align-items: center;
    justify-content: center;
    transition: background-color 0.3s;
  }
</style>

<div class="gravatar">
{#await resolveGravatar() then gravatarUrl}
  {#if gravatarUrl !== null}
  <img alt={`${user.firstName} ${user.lastName}`} src={gravatarUrl} />
  {:else}
  <div class="gravatar-intial" style="--gravatar-color: {getColor()}" >{getUserInitial()}</div>
  {/if}
{/await}
</div>
