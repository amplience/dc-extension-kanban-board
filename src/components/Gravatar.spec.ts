import { render } from '@testing-library/svelte';
import { tick } from 'svelte';
import Gravatar from './Gravatar.svelte';

jest.mock('md5', () => ({ default: jest.fn().mockReturnValue('HASH') }));

xdescribe('Gravatar', () => {
  beforeEach(() => {
    global.fetch = jest.fn(() => Promise.resolve({ ok: true } as Response));
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should render a placeholder image when no gravatar is available', async () => {
    (global.fetch as jest.Mock).mockImplementation(() =>
      Promise.resolve({ ok: false } as Response)
    );
    const { container } = render(Gravatar, {
      user: {
        firstName: 'TEST',
        lastName: 'USER',
        email: 'test-user@bigcontent.io',
      },
    });
    await tick();
    await tick();
    await tick();
    await tick();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a gravatar', async () => {
    (global.fetch as jest.Mock).mockImplementation(() =>
      Promise.resolve({ ok: true } as Response)
    );
    const { container } = render(Gravatar, {
      user: {
        firstName: 'TEST',
        lastName: 'USER',
        email: 'test-user@bigcontent.io',
      },
    });

    await tick();
    await tick();
    await tick();
    await tick();
    expect(container.firstChild).toMatchSnapshot();
  });
});
