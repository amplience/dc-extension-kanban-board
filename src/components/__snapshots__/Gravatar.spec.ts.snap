// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gravatar should render a gravatar 1`] = `
<div>
  <div
    class="gravatar svelte-17a39iv"
  >
    <img
      alt="TEST USER"
      class="svelte-17a39iv"
      src="https://www.gravatar.com/avatar/50b7748612b28db487d115f220bb77ab?d=404&s=56"
    />
    
  </div>
</div>
`;

exports[`Gravatar should render a placeholder image when no gravatar is available 1`] = `
<div>
  <div
    class="gravatar svelte-17a39iv"
  >
    <div
      class="gravatar-intial svelte-17a39iv"
      style="--gravatar-color: rgb(198,78,103);"
    >
      T
    </div>
    
  </div>
</div>
`;
